pilot/pkg/networking/core/listener.go

```diff
@@ -592,7 +592,10 @@ func buildListenerFromEntry(builder *ListenerBuilder, le *outboundListenerEntry,
 
 	if needHTTPInspector {
 		l.ListenerFilters = append(l.ListenerFilters, xdsfilters.HTTPInspector)
-		// Enable timeout only if they configure it and we have an HTTP inspector.
+	}
+
```

Comment:
if no listener filters are configured, why would envoy check for this timeout?

Reply:
I think before this change we only configured a timeout if the http inspector was added, not taking into account the TLS inspector

Reply:
Ok. I see

Reply:
No, we did not configure the timeout at all in 99.99% of cases. See the `This is really unsafe, so hopefully not used...` comment. Its off by default.

Reply:
@howardjohn Yes, currently, the timeout is only configured when an HTTPInspector exists in the listener filters (with a default value of 5 seconds, which can be customized by users, though there's a risk involved if users set this value themselves, as I understand is the point of this comment);
The [issue](https://github.com/istio/istio/issues/52850) we encountered is that when there's no HTTPInspector but a TLSInspector present in the listener filters, not setting a timeout leads to a memory leak.

Reply:
I don't believe we are setting the timeout in any case by default. there is an option to do it for http but it's off by default because it's unsafe. 

We have seen that legitimate traffic fails even with large timeouts O(5s)

Reply:
@howardjohn Oh, yes, I see my mistake now. By default, there indeed isn't any timeout set. It appears that my fix should involve adding a default timeout. Otherwise, in scenarios with high packet loss, the issue of memory leaks would remain unresolved.

It might be worth considering directly removing this configuration and adopting Envoy's default 15-second timeout, as was the behavior in Istio versions prior to 1.19.

Reply:
In my view, if a client's ClientHello message is not completed within 5 seconds, this traffic should not be considered legitimate traffic; instead, it may potentially be a part of a Denial-of-Service attack utilizing slow requests.

Reply:
I agree we shouldn't indefinitely hold the connection/memory btw. But we should make sure we do not continue on timeout, and that the timeout is large... larger than 5s. 15s maybe. 

Is there a difference in inbound/outbound/gateway?

Reply:
Note as well it's the timeout to detect the *payload* of the tcp connection matches a certain protocol... what we want is really "timeout if the connection isn't established"

Reply:
but for example, it's perfectly legitimate to open a connection, wait 60s, then send some data on it. We cannot reject that - real apps do this for legitimate reasons

Reply:
maybe we need an alternative configuration in envoy to timeout connection established

Reply:
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener_components.proto#envoy-v3-api-field-config-listener-v3-filterchain-transport-socket-connect-timeout?

Reply:
@keithmattix This configuration will not work because the transport socket connect timeout is set after the listener filters have been processed. I don't think additional timeout configurations are necessary; since listener filters require network I/O to read data, a timeout should be set when listener filters are present.

Reply:
The issue is that it's legitimate (and seen in real world) to open a connection and not send data for a long time. Not all protocols are request-response protocols. 

The issue with packet loss seems to be around connections failing to establish not establishing and then not sending data, right?

Reply:
Ah right that config is per filter chain. In that case yeah the listener filter timeout is really what we want to check. Feels like maybe istio shouldn't set it by default but allow for easier override (e.g to the user-
Defined idle timeout)

Reply:
I realized my comments about connection establishment didn't make sense since that is handed by the OS... 

In the case where there was/is no listener filters, how would you solve this? the transport socket timeout? maybe you need both then...? 


The thing is it's not safe to set the timeout for protocol detection. TLS inspector for known-tls traffic is a bit different

Reply:
Yeah if there's no listener filter, you use the transport socket timeout since there wouldn't be any listener filters to execute and you'd go straight to FCM. Protocol detection is a conundrum...but ultimately you do need an upper bound. For an edge proxy, what's the use case for a downstream opening a connection and not sending any data for longer than, say, 15s? 30s?

Reply:
Yes, inspired by the discussions above, I believe the `transport_socket_connect_timeout` should also be set, as stated in the Envoy [issue](https://github.com/envoyproxy/envoy/issues/11426) where this feature is provided:

>  The effective value of these timeouts could be reduced based on memory pressure to **increase resiliency to memory exhaustion attacks** and improving the proxy's ability to accept legitimate connections/requests.

Reply:
We have tls_inspector at edge/ingress so sounds like a listener_filter would be valid there (maybe as part of #2471?). I take your point on arbitrary protocols (not including TLS), but it certainly seems like it ought to be configurable?

Reply:
It is configurable: ProtocolDetectionTimeout

I am fine with a timeout for TLS_inspector on the edge as well for the reasons above

Reply:
Ah it's undocumented :) https://github.com/istio/api/blob/6e7441d4920eaa11dd660dbea6b69f738ec5ede3/mesh/v1alpha1/config.proto#L68. Sounds good

Reply:
Based on the above discussion, I restored the default value (0) for ProtocolDetectionTimeout and set this timeout based on the configuration when either the HTTP Inspector or the TLS Inspector is present.
It appears that we should also modify this part of the api repository to expose this configuration in the documentation: https://github.com/istio/api/blob/6e7441d4920eaa11dd660dbea6b69f738ec5ede3/mesh/v1alpha1/config.proto#L52

Reply:
I really don't think we want the "continue on filter timeout" behavior for your use case..

Reply:
I agree, it has been modified, and I have also added the handling of logic related to the gateway.

Reply:
@howardjohn If you think these changes are fine, I will further supplement the release notes.